rfcs/0085-nixos-release-stablization.md

```diff
@@ -0,0 +1,128 @@
+---
+
+# Summary
+[summary]: #summary
+
+To bring more certainty to the release cycle, add short periods where
+breaking changes are partially restricted. A list of Release Critical
+Packages is defined. Also, move most release stablization work from
+the `release` branch to the `master` branch to reduce backports.
+
+# Motivation
+[motivation]: #motivation
+
+We want to release on time without herculian efforts. Limited
+restrictions to merging breaking changes should avoid the common
+practice of forcing many, largely-untested, commits into master before
+branch-off.
+
+This happened for the 20.09 release in which ~640 staging
+commits were pushed through in the days leading up to the branch off
+([#95492](https://github.com/NixOS/nixpkgs/pull/95492),
+[#96280](https://github.com/NixOS/nixpkgs/pull/96280),
+[#96437](https://github.com/NixOS/nixpkgs/pull/96437),
+[#97146](https://github.com/NixOS/nixpkgs/pull/97146)). The last PR included a
+release blocker fix (introduced in [one of the mentioned
+PRs](https://github.com/NixOS/nixpkgs/pull/96437)) which delayed the
+branch-off from 4 Sept 2020 to 7 Sept 2020, and ZHF had to be delayed
+another day to get a semi-accurate hydra build status. The systemd
+bump was also largely responsible delaying the release until after the
+scheduled release date where basic login features in plasma [weren't
+restored until Oct 5th, a week after the planned release
+date](https://github.com/NixOS/nixpkgs/pull/99629) and a backlog of
+other plasma related issues [weren't resolved until Oct 19th, almost a
+month after the planned release
+date](https://github.com/NixOS/nixpkgs/pull/101078).  This is not to
+discredit any individual, but demonstrate that certain packages have
+many dimensions of compatibility which need a longer time to stabilize
+on unstable before being included in a release.
+
+
+# Definitions
+[definitions]: #definition
+- Breaking change
+  - A change that is likely to break downstream.
+  - For projects following Semantic Versioning, disallow major version bumps.
+- Restricted
+  - Breaking changes are disallowed.
+  - Breaking changes for security or critical bugs are allowed. (Patching is preferred over updates.)
+  - New packages and NixOS modules are welcome.
+- Unrestricted
+  - Breaking changes are allowed.
+  - New packages and NixOS modules are welcome.
+- ZHF: Zero Hydra Failures
+  - Period in which packages are stabilized in preparation of a release.
+
+# Changes
+
+## Release Critical Packages
+
+A list of Release Critical Packages will be maintained in the nixpkgs
+manual.  Release Critical Packages are important packages which have
+many dimensions of build or runtime behavior. Generally these will be
+packages which are a part of `stdenv`, or bootstrapping a system
+(e.g. systemd). The Release Managers are empowered to decide which
+packages are Release Critical Packages without additional RFCs.
+
+Initially these will be `stdenv.cc`,`binutils`, and `systemd`.
```

Comment:
Still not very happy about calling out any specific packages in here. The paragraph above is explicit enough IMHO.

Reply:
Could you go into more detail about why you are opposed to this? I feel like it gives important context for future Release Managers when they decide what should be a Release Critical Package.

Reply:
Only three were mentioned, and each of them have significant build and runtime behavior which needs to be stabilized on master before branch-off.

stdenv also includes other packages such as `stdenv.shellPackage`, `stdenv.bootstrapTools`, and a few others. But I guess you could argue those are highly to cause stabilization pain if bumped as well. I'm open to changing the line to just read as
```suggestion
Initially these will be all packages included in `stdenv` and `systemd`.
```

Reply:
A refactoring of nixpkgs or stdenv should not make this RFC obsolete. You can list those examples in an advisory section for release manager but including them in part of the "law" just ensures that it will be outdated at some point.

Reply:
As an alternative should this just be linked to to the release maintainer wiki? https://github.com/NixOS/release-wiki

Reply:
I don't see systemd, gcc, binutils, or bootstrapping packages to be obsolete anytime soon.

I believe I also mentioned that the release retrospectives will be the appropriate time to revisit these lists and edit them. However, I think think it's safe to say that gcc, binutils, and systemd will be "high-risk" packages for the foreseeable future; and that bumping them immediately before ZHF cause more issues than it solves.

Reply:
@jonringer Do you prefer release-wiki to nixpkgs manual for storing a list of RCPs?

I thought about @andir's thing more, and I now do not object to removing the "Initially.." sentence.

Reply:
I would prefer the release-wiki. But either is okay.

Reply:
@jonringer I think you should change it to be the release wiki. It makes more sense.

Reply:
> so in best case they [the "Release Critical Package"] never break

I think the issue is less these packages themselves "breaking" (in the sense of not being buildable anymore or not functioning correctly when taken for themselves), but for changes to them to break many depending packages.

Reply:
List of release critical packages will now be curated on the release wiki